marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/CreateGranularPrivilegesCommand.java

```diff
@@ -95,160 +102,196 @@ public Integer getUndoSortOrder() {
-
-        list.add(buildPrivilege(client, "admin-database-index-" + stagingDbName, "http://marklogic.com/xdmp/privileges/admin/database/index/$$database-id(" + stagingDbName + ")",
-            "STAGING-index-editor", existingPrivilegeNames, developerRole, hubCentralEntityModelWriterRole));
-        list.add(buildPrivilege(client, "admin-database-index-" + finalDbName, "http://marklogic.com/xdmp/privileges/admin/database/index/$$database-id(" + finalDbName + ")",
-            "FINAL-index-editor", existingPrivilegeNames, developerRole, hubCentralEntityModelWriterRole));
-        list.add(buildPrivilege(client, "admin-database-index-" + jobsDbName, "http://marklogic.com/xdmp/privileges/admin/database/index/$$database-id(" + jobsDbName + ")",
-            "JOBS-index-editor", existingPrivilegeNames, developerRole));
-
-        return list;
+        return granularPrivilegeMap;
     }
 
     /**
-     * @param client
-     * @param name                   The name of the privilege to create if a privilege with the given dhsName does not exist. This name
-     *                               is not the same as the dhsName, as the initial goal was to have a consistent naming convention for
-     *                               all granular privileges, and the DHS privilege names are not consistent with that convention.
-     * @param action
-     * @param dhsName
-     * @param existingPrivilegeNames
-     * @param rolesToAdd
-     * @return
+     * Save each of the given privileges. For each key in the map - where the key is expected to be an action with a
+     * resource ID in it - we check to see if an existing privilege has the same action. If so, then the roles in the
+     * granular privilege are added to that existing privilege. This ensures we never cause an error by trying to create
+     * a privilege with the same action as an existing one. This is crucial for DHS, as the DHS config will create some
+     * of the same granular privileges that DHF needs to create (but with a different name).
+     *
+     * @param manageClient
+     * @param granularPrivileges
      */
-    protected Privilege buildPrivilege(ManageClient client, String name, String action, String dhsName, List<String> existingPrivilegeNames, String... rolesToAdd) {
-        Privilege p;
-        if (existingPrivilegeNames.contains(dhsName)) {
-            final String json = new PrivilegeManager(client).getAsJson(dhsName, "kind", "execute");
-            p = new DefaultResourceMapper(new API(client)).readResource(json, Privilege.class);
-        } else {
-            p = new Privilege(null, name);
-            p.setKind("execute");
-            p.setAction(action);
-        }
-        for (String role : rolesToAdd) {
-            p.addRole(role);
-        }
-        return p;
+    protected void saveGranularPrivileges(ManageClient manageClient, Map<String, Privilege> granularPrivileges) {
```

Comment:
This is the method that every in-memory privilege goes through, it's where the new logic is.